### PR TITLE
Added tests and coverage for CLI

### PR DESCRIPTION
**Pull request by vishav1771**
## Description

This PR implements the first two tasks of issue #297. 

- Added tests and coverage for `__main__.py` , and reached the coverage of about 80%.

## Motivation and Context
#297 

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

Example for a changelog entry: `Fix issue with launching rockets to the moon (#101, #212).`

### CODE REVIEW

**Comment by csadorf** [comment]
This looks really good! Thank you for pushing work-in-progress.

There appears to be some kind of unrelated issue with tests on py36. I've tried to investigate a little bit, but so far am not sure what is going on. I suggest to ignore that for now.

**Comment by vishav1771** [comment]
@csadorf I have added a test for `export` option. Please have a look at the CI error.
I think this can be removed by adding `os.mkdir`. Is that okay? (because we don't need that in `export_to`)

**Comment by csadorf** [comment]
> @csadorf I have added a test for `export` option. Please have a look at the CI error.
> I think this can be removed by adding `os.mkdir`. Is that okay? (because we don't need that in `export_to`)

@vishav1771 It should not be necessary to create the directory that we export to, because it is created automatically. So I am not sure what is causing the issue.

Things to check:
- [ ] Does the `/tmp` directory exist and do we have write access to it?
- [ ] Is there some kind of issue with finding the `python` executable?

**Comment by csadorf** [comment]
@vishav1771 Impressive progress on the testing front! You can use [this page](https://codecov.io/gh/glotzerlab/signac/src/dc12796366edcd93d89c1bd023a68f2ceb3f027f/signac/__main__.py) provided by codecov to identify missing hotspots. As you know, the goal is to reach 80% coverage, but please let us know if you are running into trouble reaching that.

**Comment by bdice** [comment]
@vishav1771 I looked at the lines that are missing coverage. A lot of the coverage misses have to do with deprecated modules, which would probably be a pain to test. I would focus on non-deprecated code paths. Here are a few easy tests that should be near the top of the `__main__` tests. (I haven't checked if these tests pass but they're fairly close.)

```python
    def test_print_usage(self):
        out = self.call('python -m signac'.split())
        assert 'usage:' in out

    def test_version(self):
        out = self.call('python -m signac --version'.split())
        assert 'signac {}'.format(signac.__version__) in out

    def test_help(self):
        out = self.call('python -m signac --help'.split())
        assert 'positional arguments:' in out
        assert 'optional arguments:' in out

**Comment by vishav1771** [comment]
@csadorf @bdice Thanks for the review, and providing with the necessary changes. I'll try to reach the required 80% coverage mark.
While writing test for  `config verify`, I was getting the error for `signac config verify `.
```
Error: validate() got an unexpected keyword argument 'skip_missing'
Traceback (most recent call last):
  File "C:\Users\visha\AppData\Local\Continuum\anaconda3\Scripts\signac-script.py", line 11, in <module>
    load_entry_point('signac==1.3.0', 'console_scripts', 'signac')()
  File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\__main__.py", line 1835, in main
  File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\__main__.py", line 809, in main_config_verify
  File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\__main__.py", line 722, in verify_config
  File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\common\config.py", line 133, in verify
TypeError: validate() got an unexpected keyword argument 'skip_missing'
```
Can you please verify this behaviour. Also, can you provide me with examples where the verification of config file fails.

**Comment by csadorf** [comment]
> @csadorf @bdice Thanks for the review, and providing with the necessary changes. I'll try to reach the required 80% coverage mark.
> While writing test for `config verify`, I was getting the error for `signac config verify `.
> 
> ```
> Error: validate() got an unexpected keyword argument 'skip_missing'
> Traceback (most recent call last):
>   File "C:\Users\visha\AppData\Local\Continuum\anaconda3\Scripts\signac-script.py", line 11, in <module>
>     load_entry_point('signac==1.3.0', 'console_scripts', 'signac')()
>   File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\__main__.py", line 1835, in main
>   File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\__main__.py", line 809, in main_config_verify
>   File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\__main__.py", line 722, in verify_config
>   File "C:\Users\visha\AppData\Local\Continuum\anaconda3\lib\site-packages\signac-1.3.0-py3.7.egg\signac\common\config.py", line 133, in verify
> TypeError: validate() got an unexpected keyword argument 'skip_missing'
> ```
> 
> Can you please verify this behaviour. Also, can you provide me with examples where the verification of config file fails.

Looks like you identified a completely unrelated bug. That's great, because it shows how important testing is and we were able to detect this bug because of your hard work.

I've looked into this locally and found that #302 should resolve the issue. Once #302 is merged, we can rebase this branch and move forward.

**Comment by vishav1771** [comment]
Thanks, @csadorf , I'll continue with this PR after #302 is merged.
Also, while playing around, I think I found another unrelated bug.
While executing `signac sync C:\Users\visha\Desktop\test C:\Users\visha\Desktop\workspace -w`
I am getting the error: 
```
Error: 'schema_version'
```
Can you please verify this?

**Comment by bdice** [comment]
I can see the error as well. This appears to be caused during the explicit instantiation of the Project config here, which doesn't include the mandatory `schema_version` key: https://github.com/glotzerlab/signac/blob/fa2d8c40dd5f0d132dfc9c67a2570277c264f8af/signac/__main__.py#L474-L478

I suppose we have to assume a schema version for the workspace (since there's no project configuration to read from) or disallow the option to synchronize directly to the workspace.

**Comment by vishav1771** [comment]
I think we can pass  ` _ignore_schema_version=True` while creating Project to remove the error.

https://github.com/glotzerlab/signac/blob/fa2d8c40dd5f0d132dfc9c67a2570277c264f8af/signac/__main__.py#L474-L478

**Comment by csadorf** [comment]
> I think we can pass ` _ignore_schema_version=True` while creating Project to remove the error.
> 
> https://github.com/glotzerlab/signac/blob/fa2d8c40dd5f0d132dfc9c67a2570277c264f8af/signac/__main__.py#L474-L478

I would advocate for using `_ignore_schema_version=True` as a work-around, but generally deprecating this option.

@bdice The problem further shows why allowing workspaces outside project root directories is not a good idea. In this case there is absolutely no way for us to know what schema that particular workspace is possibly maintained with. The relationship between project and workspace should be bijective.

**Comment by csadorf** [comment]
> Thanks, @csadorf , I'll continue with this PR after #302 is merged.

You can probably continue working on this PR in parallel, you just need to ignore that particular code section.

**Comment by bdice** [comment]
@vishav1771 #302 has been merged.

File: tests/test_shell.py [hunk 1]
```diff
@@ -113,9 +130,13 @@ def test_statepoint(self):
         self.call('python -m signac init my_project'.split())
         self.call(['python', '-m', 'signac', 'job', '--create', '{"a": 0}'])
         project = signac.Project()
-        for job in project:
-            sp = self.call('python -m signac statepoint {}'.format(job).split())
-            assert project.open_job(json.loads(sp)) == job
+        job = project.open_job({'a': 0})
+        sp = self.call('python -m signac statepoint {}'.format(job).split())
```
Comment (bdice) [inline tests/test_shell.py]: Let's make it explicit that we use the job id here.
```suggestion
        sp = self.call('python -m signac statepoint {}'.format(job.id).split())
```

File: tests/test_shell.py [hunk 2]
```diff
@@ -113,9 +130,13 @@ def test_statepoint(self):
...
+        assert project.open_job(json.loads(sp)) == job
+        sp = self.call('python -m signac statepoint'.split())
+        assert project.open_job(json.loads(sp)) == job
+        sp = self.call('python -m signac statepoint --pretty'.split())
+        assert "{'a': 0}" in sp
```
Comment (bdice) [inline tests/test_shell.py]: I found some inconsistent behavior. In general, I believe the signac CLI should always have JSON-compatible output where possible. In a test project, I ran the following:
```bash
$ signac statepoint
{"a": {"b": 0, "c": 0}}
$ signac statepoint --pretty
{'a': {'b': 0, 'c': 0}}
```
The `--pretty` output is not JSON-compatible because JSON requires double quotes, not single quotes. @csadorf Is this something we've thought about before? I don't know how `--pretty` is implemented.

File: tests/test_shell.py [hunk 3]
```diff
@@ -129,13 +150,35 @@ def test_index(self):
...
         project.open_job({'a': 0}).document['b'] = 0
         doc = json.loads(self.call('python -m signac index'.split()))
         assert 'statepoint' in doc
         assert doc['statepoint'] == {'a': 0}
         assert 'b' in doc
         assert doc['b'] == 0
 
+    def test_document(self):
+        self.call('python -m signac init my_project'.split())
+        self.call('python -m signac project --access'.split())
```
Comment (bdice) [inline tests/test_shell.py]: This `signac project --access` command doesn't seem related to testing job documents, so it probably goes somewhere else. The generated `signac_access.py` also appears to use deprecated functionality. @csadorf Do you know if this is supposed to be deprecated overall? I'm not familiar with this part of the codebase, I didn't know this feature for creating "access modules" existed and am not sure whether it still serves a purpose.

File: tests/test_shell.py [hunk 4]
```diff
@@ -129,13 +150,35 @@ def test_index(self):
...
+        job_a.init()
+        assert len(project) == 1
+        job_a.document['data'] = 4
+        doc = json.loads(self.call('python -m signac document'.split()))
+        assert 'data' in doc
+        assert doc['data'] == 4
+        doc = json.loads(self.call('python -m signac document {}'.format(job_a.id).split()))
+        assert 'data' in doc
+        assert doc['data'] == 4
+        out = self.call('python -m signac document --pretty'.split())
```
Comment (bdice) [inline tests/test_shell.py]: Same issue as above with JSON compatibility and single quotes with `--pretty`, double quotes without `--pretty`.

File: tests/test_shell.py [hunk 5]
```diff
@@ -188,6 +251,68 @@ def test_find(self):
...
 
+    def test_diff(self):
+        self.call('python -m signac init ProjectA'.split())
+        project = signac.Project()
+        job_a = project.open_job({"a": 0, "b": 1})
+        job_a.init()
+        job_b = project.open_job({"a": 0, "b": 0})
+        job_b.init()
+        out = self.call('python -m signac diff {} {}'.format(job_a.id, job_b.id).split())
+        expected = [str(job_a.id), "{'b': 1}", str(job_b.id), "{'b': 0}"]
```
Comment (bdice) [inline tests/test_shell.py]: Same potential issue with single/double quotes.

File: tests/test_shell.py [hunk 6]
```diff
@@ -199,13 +324,234 @@ def test_remove(self):
...
+                      .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+
+        self.call('python -m signac sync {} {} --no-key'
+                  .format(os.path.join(self.tmpdir.name, 'b'), self.tmpdir.name).split())
+        assert job_dst.document != job_src.document
+        assert job_dst.document['a'] != job_src.document['a']
+        assert job_dst.document['nested'] != job_src.document['nested']
+
+        reset()     # only sync a
```
Comment (bdice) [inline tests/test_shell.py]: Split these comments onto their own lines, above `reset()`. Also add a comment explaining what is happening for all the options (`--no-key`, `--all-key`, etc.).

File: tests/test_shell.py [hunk 7]
```diff
@@ -199,13 +324,234 @@ def test_remove(self):
...
+
+    def test_sync_file(self):
+        self.call('python -m signac init ProjectA'.split())
+        project_a = signac.Project()
+        project_b = signac.init_project('ProjectB', os.path.join(self.tmpdir.name, 'b'))
+        job_src = project_a.open_job({'a': 0})
+        job_dst = project_b.open_job({'a': 0})
+        for i, job in enumerate([job_src, job_dst]):
+            with job:
+                with open('test', 'w') as file:
```
Comment (bdice) [inline tests/test_shell.py]: Could be simplified so that `with job` isn't needed: `with open(job.fn('test'), 'w') as file:`

File: tests/test_shell.py [hunk 8]
```diff
@@ -199,13 +324,234 @@ def test_remove(self):
...
+        assert "Import failed" in err
+        self.call("python -m signac import {} --sync".format(prefix_data).split(), error=True)
+        assert len(project_a) == 4
+        assert 'a' in job_dst.document
+        assert job_dst.document['a'] == 0
+        out = self.call(
+            'python -m signac import {} --sync-interactive'.format(prefix_data),
+            'print(str(tmp_project), len(tmp_project)); exit()', shell=True)
+        assert 'ProjectA' in out
+        assert '4' in out
```
Comment (bdice) [inline tests/test_shell.py]: Combine if possible.
```suggestion
        assert 'ProjectA 4' in out
```

File: tests/test_shell.py [hunk 9]
```diff
@@ -215,7 +561,15 @@ def test_shell(self):
             'print(str(project), job, len(list(jobs))); exit()', shell=True)
         assert out.strip() == '>>> {} None {}'.format(project, len(project))
 
+        out = self.call(
+            'python -m signac shell -c print(str(project),len(list(jobs)))'
+            .split())
```
Comment (bdice) [inline tests/test_shell.py]: This is a little bit of a hack, it looks like you're avoiding spaces in the `-c` argument to prevent splitting that token. I suggest either specifying this as input or using another way to tokenize manually like:
```suggestion
        cmd = 'python -m signac shell -c'.split() + ['print(str(project), len(list(jobs)))']
        out = self.call(cmd)
```

File: tests/test_shell.py [hunk 10]
```diff
@@ -244,9 +598,97 @@ def test_shell_with_jobs_and_selection_only_one_job(self):
...
+        assert out.split(os.linesep) == expected
+
+        out = self.call('python -m signac config --global show'.split()).strip()
+        cfg = config.read_config_file(config.FN_CONFIG)
+        expected = config.Config(cfg).write()
+        assert out.split(os.linesep) == expected
+
+    def test_config_set(self):
+        self.call('python -m signac init my_project'.split())
+        self.call('python -m signac config  set a b'.split())
```
Comment (bdice) [inline tests/test_shell.py]:
```suggestion
        self.call('python -m signac config set a b'.split())
```

File: tests/test_shell.py (hunk 2 above)
Comment (csadorf) [reply tests/test_shell.py]: We use `pprint` to implement pretty. We should probably have used some JSON prettifying instead. Feel free to classify as bug and fix without deprecation cycle.

File: tests/test_shell.py (hunk 3 above)
Comment (csadorf) [reply tests/test_shell.py]: This is deprecated.

File: tests/test_shell.py (hunk 4 above)
Comment (csadorf) [reply tests/test_shell.py]: Same comment as above.

File: tests/test_shell.py (hunk 2 above)
Comment (vyasr) [reply tests/test_shell.py]: Switching from pprint to something like `json.dumps` would also give us something closer to the output that @phmalek requested recently on Gitter

File: tests/test_shell.py (hunk 3 above)
Comment (bdice) [reply tests/test_shell.py]: I created #303 to address this.

File: tests/test_shell.py (hunk 2 above)
Comment (bdice) [reply tests/test_shell.py]: While `json.dumps` offers an `indent`, there is no way to have the equivalent of pprint's `depth` flag with `json.dumps`:

```python
>>> from pprint import pprint
>>> pprint({'a': {'very': {'deep': 'dictionary'}}})
{'a': {'very': {'deep': 'dictionary'}}}
>>> pprint({'a': {'very': {'deep': 'dictionary'}}}, depth=2)
{'a': {'very': {...}}}
```
@csadorf @vyasr Is this "depth" a hard requirement for pretty output?

Comment (csadorf) [reply tests/test_shell.py]: I did find that kind of important in my interactive work, yes.

Comment (bdice) [reply tests/test_shell.py]: In that case, I recommend to make no changes to existing behavior. If "pretty" output was just about indenting it, then that can be solved with `json.dumps`. Since there's not an easy way to solve this, let's leave the current behavior as-is. If you agree, we can resolve the conversations about pretty output.

File: tests/test_shell.py (hunk 8 above)
Comment (vishav1771) [reply tests/test_shell.py]: I don't we can combine both of them as `tmp_project.id` is combination of ` project.id, str(uuid.uuid4()) `

Comment (bdice) [reply tests/test_shell.py]: Oh, okay. Nevermind. I'll resolve this.

File: requirements-dev.txt [hunk 11]
```diff
@@ -8,3 +8,5 @@ click>=7.0
 ruamel.yaml>=0.15.89
 pytest==5.3.4
 pytest-subtests==0.3.0
+pytest-cov==2.8.1
+pymongo==3.10.1
```
Comment (csadorf) [inline requirements-dev.txt]: We are actually trying to move away from the optional pymongo integration, but I guess we can always drop that at some later point.

File: tests/test_shell.py [hunk 12]
```diff
@@ -64,8 +65,24 @@ def call(self, command, input=None, shell=False):
...
         if p.returncode != 0:
             raise ExitCodeError("STDOUT='{}' STDERR='{}'".format(out, err))
+        if error:
+            return err.decode()
         return out.decode()
 
+    def test_print_usage(self):
+        with pytest.raises(ExitCodeError):
+            out = self.call('python -m signac'.split())
+            assert 'usage:' in out
```
Comment (csadorf) [inline tests/test_shell.py]: Considering that `pytest.raises(..)` is implying that we assume that the `call()` is raising an `ExitCodeError`, would this assert never be tested?

File: tests/test_shell.py [hunk 13]
```diff
@@ -113,9 +130,13 @@ def test_statepoint(self):
         self.call('python -m signac init my_project'.split())
         self.call(['python', '-m', 'signac', 'job', '--create', '{"a": 0}'])
         project = signac.Project()
-        for job in project:
-            sp = self.call('python -m signac statepoint {}'.format(job).split())
-            assert project.open_job(json.loads(sp)) == job
+        job = project.open_job({'a': 0})
+        sp = self.call('python -m signac statepoint {}'.format(job.id).split())
+        assert project.open_job(json.loads(sp)) == job
```
Comment (csadorf) [inline tests/test_shell.py]: Can we add an `assert len(project) == 1` below this line, I was initially confused by the following asserts.

File: tests/test_shell.py [hunk 14]
```diff
@@ -188,6 +250,68 @@ def test_find(self):
...
+    def test_clone(self):
+        self.call('python -m signac init ProjectA'.split())
+        project_a = signac.Project()
+        project_b = signac.init_project('ProjectB', os.path.join(self.tmpdir.name, 'b'))
+        job = project_a.open_job({'a': 0})
+        job.init()
+        assert len(project_a) == 1
+        assert len(project_b) == 0
+        with pytest.raises(ExitCodeError):
+            self.call("python -m signac clone {} 9bfd29df07674bc5"
```
Comment (csadorf) [inline tests/test_shell.py]: Can we add a comment here about the meaning of that hard-coded id? Is it expected to already exist?

File: tests/test_shell.py [hunk 15]
```diff
@@ -188,6 +250,68 @@ def test_find(self):
...
+    def test_move(self):
+        self.call('python -m signac init ProjectA'.split())
+        project_a = signac.Project()
+        project_b = signac.init_project('ProjectB', os.path.join(self.tmpdir.name, 'b'))
+        job = project_a.open_job({'a': 0})
+        job.init()
+        assert len(project_a) == 1
+        assert len(project_b) == 0
+        with pytest.raises(ExitCodeError):
+            self.call("python -m signac move {} 9bfd29df07674bc5"
```
Comment (csadorf) [inline tests/test_shell.py]: Same comment as above.

File: tests/test_shell.py [hunk 16]
```diff
@@ -199,14 +323,233 @@ def test_remove(self):
         assert job_to_remove in project
         assert job_to_remove.doc.a == 0
         assert len(job_to_remove.doc) == 1
-        with pytest.deprecated_call():
```
Comment (csadorf) [inline tests/test_shell.py]: Are those calls not considered deprecated anymore?

File: tests/test_shell.py [hunk 17]
```diff
@@ -199,14 +323,233 @@ def test_remove(self):
...
-            self.call('python -m signac rm --clear {}'.format(job_to_remove.get_id()).split())
+        self.call('python -m signac rm --clear {}'.format(job_to_remove.id).split())
         assert job_to_remove in project
         assert len(job_to_remove.doc) == 0
-        with pytest.deprecated_call():
-            self.call('python -m signac rm {}'.format(job_to_remove.get_id()).split())
+        self.call('python -m signac -v rm {}'.format(job_to_remove.id).split())
+        assert job_to_remove not in project
+        with pytest.raises(ExitCodeError):
```
Comment (csadorf) [inline tests/test_shell.py]: I think some comment (here and in similar locations elsewhere) as to why certain calls are expected to fail would be very helpful.

File: tests/test_shell.py (hunk 16 above)
Comment (vishav1771) [reply tests/test_shell.py]: The deprecated call was because of `get_id` function. I removed it so that it's not needed anymore.

File: requirements-dev.txt (hunk 11 above)
Comment (vishav1771) [reply requirements-dev.txt]: `pymongo` is required for the `config host` command. This can be removed after the its deprecation cycle.

File: tests/test_shell.py (hunk 12 above)
Comment (vishav1771) [reply tests/test_shell.py]: Sorry, I didn't realize that earlier. will change.

File: tests/test_shell.py [hunk 18]
```diff
@@ -269,9 +273,6 @@ def test_clone(self):
         job.init()
         assert len(project_a) == 1
         assert len(project_b) == 0
-        with pytest.raises(ExitCodeError):
-            self.call("python -m signac clone {} 9bfd29df07674bc5"
-                      .format(os.path.join(self.tmpdir.name, 'b')).split())
```
Comment (csadorf) [inline tests/test_shell.py]: Why was this removed? This seems like the negative test corresponding to the positive test below.

Comment (vishav1771) [reply tests/test_shell.py]: This test was not removed, I just changed its position. I was trying to arrange the test in a certain order (like first do all positive tests then negative tests).

**Comment by csadorf** [comment]
@vyasr Merge at your discretion.
